### PR TITLE
#163518690 User should be able to rate articles

### PR DESCRIPTION
**Pull request by espoirMur**
#### What does this PR do?
This PR adds the functionality of creating, updating, fetching rating on a given article

#### Description of Task to be completed?
- add create rating endpoint 
- add get all rating endpoint
- add update rating endpoint
- secure all endpoints with Jwt
- restructure the test suite for the application

#### How should this be manually tested?
* Open Postman client
* Consume the API:
    * `post: /api/v1/article/:slug/rating` supply the user 's token,and number of rating.
    * `put: /api/v1/article/:slug/rating/:ratingId` supply the user 's token, and the updated rate.
    * `get: api/v1/article/:slug/ratingId` supply the user 's token.

### CODE REVIEW

**Review by YvesIraguha — APPROVED** [review]

**Review by YvesIraguha — COMMENTED** [review]
Please, resolve the conflicts in the conflicting files. 
`controllers/auth/loginController.js`
`controllers/users.js`
`index.js`
`middlewares/checkAuth.js`
`middlewares/passport.js`
`models/Article.js`
`routes/articles.js`
`routes/index.js`
`sample.env`
`swagger.json`
`tests/integration/article.test.js`
`tests/integration/login.test.js`
`tests/integration/signup.test.js`

File: controllers/articles.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+              status: CREATED,
+              message: "Article created",
+              article: article.dataValues
+            })
+          : res
+              .status(NOT_FOUND)
+              .json({ status: NOT_FOUND, message: "Please consider logging in" });
+      }
+    } catch (error) {
+      // console.log("DATABASE ERROR:", error);
```
Comment (puritybirir) [inline controllers/articles.js]: Nitpick: no-console

File: index.js [hunk 2]
```diff
@@ -41,43 +46,32 @@ if (!isProduction) {
 
 app.use("/api-docs", swaggerUi.serve, swaggerUi.setup(swaggerDocument));
 app.use("/api/v1/", routers);
-
+app.use(joiValidator());
 // catch 404 and forward to error handler
```
Comment (puritybirir) [inline index.js]: Remove this comment

File: controllers/ratings.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    const articleId = req.params.id;
+    const userId = req.user.id;
+
+    joi.validate({ userId, articleId }, ratingAll, (err, value) => {
+      if (err) {
+        next(err);
+      } else {
+        // Ps: user raw query because using findOne was causing errors because of table aming
+        sequelize
+          .query('SELECT * FROM "ratings"  WHERE "articleId" =:articleId', {
```
Comment (puritybirir) [inline controllers/ratings.js]: Use sequelize and not raw SQL queries

File: middlewares/passport.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+const { Strategy, ExtractJwt } = passportJwt;
+const { User } = models;
+const options = {
+  jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+  secretOrKey: process.env.SECRET_KEY
+};
+
+const { INTERNAL_SERVER_ERROR, UNAUTHORIZED } = constants.statusCode;
+/**
+ * @description - The middleware to check for authentication. It checks on every
```
Comment (puritybirir) [inline middlewares/passport.js]: Remain with only the first part of the description

File: routes/ratings.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+const ratingRouters = Router();
+ratingRouters.use((req, res, next) => {
+  passport.authenticate("jwt", { session: false }, (err, user, info) => {
+    if (!user) {
+      return res.status(401).json({
+        message: "Please provide a token to perform this action"
+      });
+    }
+    // pass the user to the next middleware
```
Comment (puritybirir) [inline routes/ratings.js]: Remove inline comment

File: controllers/auth/loginController.js [hunk 6]
```diff
@@ -2,7 +2,9 @@ import models from "../../models";
 import jwt from "jsonwebtoken";
 import validate from "../../helpers/validators/login.validation";
 import bcrypt from "bcrypt";
+import dotenv from "dotenv";
 
+dotenv.config();
 const { User } = models;
 /** Define class for handling user login */
```
Comment (abayo-luc) [inline controllers/auth/loginController.js]: use JsDoc comment

File: controllers/ratings.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          .then(result => {
+            if (result._options.isNewRecord) {
+              res.status(201).json({
+                message: "Post rated successfully",
+                rating: result.dataValues
+              });
+            }
+          })
+          .catch(error => {
+            // check if the error is related to the post id not find in table
```
Comment (abayo-luc) [inline controllers/ratings.js]: Remove these comments

File: controllers/ratings.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
+/* eslint-disable no-underscore-dangle */
+import joi from "joi";
+import models from "../models";
+import { ratingOne, ratingAll } from "../helpers/ratingValidation";
+
+const { Rating, sequelize } = models;
+
+class RatingController {
+  static create(req, res, next) {
+    const articleId = req.params.id;
```
Comment (abayo-luc) [inline controllers/ratings.js]: use ```slug``` instead of ```id```

File: controllers/ratings.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    const userId = req.user.id;
+    const { rating } = req.body;
+    joi.validate({ articleId, userId, rating }, ratingOne, (err, value) => {
+      if (err) {
+        next(err);
+      } else {
+        Rating.create({ articleId, userId, rating })
```
Comment (abayo-luc) [inline controllers/ratings.js]: are you saving ```rating``` directly from the request body? I think you should be saving an average or something?

File: controllers/ratings.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            res.status(200).json({ ratings: result });
+          })
+          .catch(error => {
+            res.status(500).json({ message: "Please Try again later" });
+          });
+      }
+    });
+  }
+
+  static update(req, res, next) {
```
Comment (abayo-luc) [inline controllers/ratings.js]: I don't this updating ```rating``` is necessary.

**Review by abayo-luc — CHANGES_REQUESTED** [review]

File: controllers/ratingsController.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
+import joi from "joi";
+import models from "../models";
+import { ratingOne, ratingAll } from "../helpers/ratingValidation";
+import constants from "../helpers/constants";
+
+const { Rating, sequelize } = models;
+const { NOT_FOUND, OK, INTERNAL_SERVER_ERROR } = constants.statusCode;
+class RatingController {
```
Comment (puritybirir) [inline controllers/ratingsController.js:8]: Add space here

File: tests/integration/rating.test.js [hunk 12]
```diff
@@ -0,0 +1,129 @@
+import dotenv from "dotenv";
```
Comment (puritybirir) [inline tests/integration/rating.test.js:1]: Use the format for testing used by the rest of the team

**Review by puritybirir — CHANGES_REQUESTED** [review]

File: controllers/ratingsController.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    const userId = req.user.id;
+    const { rating } = req.body;
+    const { article } = req;
+    joi.validate({ slug, userId, rating }, ratingOne, (err, _value) => {
+      if (err) {
+        next(err);
+      } else {
+        Rating.create({ slug, userId, rating, articleId: article.id })
+          .then(result => {
+            /* eslint-disable no-underscore-dangle */
```
Comment (abayo-luc) [inline controllers/ratingsController.js:28]: Remove this comment!

File: controllers/ratingsController.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        next(err);
+      } else {
+        try {
+          const results = await Rating.findAll({
+            raw: true,
+            where: { articleId: article.id }
+          });
+          const averageRating = await Rating.findAll({
+            attributes: [
+              [sequelize.fn("AVG", sequelize.col("rating")), "averageRating"]
```
Comment (abayo-luc) [inline controllers/ratingsController.js:82]: What does this do?

**Review by abayo-luc — CHANGES_REQUESTED** [review]
- Improve your test coverage
- Rebase develop
- Remove comments in the codebase. If it will take you to add some configuration to .eslintric, feel free and communicate to the team. 
- I wasn't able to rate an article, I don't know why it kept giving me error ```The page you are looking for cannot be found```, yet I am sure I have created an article.

**Review by abayo-luc — APPROVED** [review]

**Review by puritybirir — APPROVED** [review]

**Comment by espoirMur** [comment]
> * Improve your test coverage
> * Rebase develop
> * Remove comments in the codebase. If it will take you to add some configuration to .eslintric, feel free and communicate to the team.
> * I wasn't able to rate an article, I don't know why it kept giving me error `The page you are looking for cannot be found`, yet I am sure I have created an article.

**Comment by espoirMur** [comment]
Thanks, @abayo-luc  let me do that...

**Comment by espoirMur** [comment]
> * Improve your test coverage
> * Rebase develop
> * Remove comments in the codebase. If it will take you to add some configuration to .eslintric, feel free and communicate to the team.
> * I wasn't able to rate an article, I don't know why it kept giving me the error `The page you are looking for cannot be found`, yet I am sure I have created an article.

Thanks, @abayo-luc  for that, Let me fix it...

**Pull request merged**